addons/mail_group/models/mail_group.py

```diff
@@ -0,0 +1,319 @@
+    def message_update(self, message_dict=None):
+        """Add the method to make the mail gateway flow work with this model."""
+        return
+
+    def message_post(self, **kwargs):
+        self.ensure_one()
+        # First create the <mail.message>
+        Mailthread = self.env['mail.thread']
+        values = dict((key, val) for key, val in kwargs.items() if key in self.env['mail.message']._fields)
+        author_id, email_from = Mailthread._message_compute_author(kwargs.get('author_id'), kwargs.get('email_from'), raise_exception=True)
+        values.update({
+            'author_id': author_id,
+            'email_from': email_from,
+            'model': self._name,
+            'res_id': self.id,
+            'partner_ids': [],
+        })
+        attachments = kwargs.get('attachments') or []
+        attachment_ids = kwargs.get('attachment_ids') or []
+        attachement_values = Mailthread._message_post_process_attachments(attachments, attachment_ids, values)
+        values.update(attachement_values)
+        mail_message = Mailthread._message_create(values)
+
+        # Create the group message associated
+        group_message = self.env['mail.group.message'].create({
+            'mail_group_id': self.id,
+            'mail_message_id': mail_message.id,
+        })
+
+        # Check the moderation rule to determine if we should accept or reject the email
+        email_normalized = email_normalize(kwargs.get('email_from'))
+        moderation = self.env['mail.group.moderation'].search([
+            ('mail_group_id', '=', self.id),
+            ('email', '=', email_normalized),
+        ], limit=1)
+
+        if not self.moderation or (moderation and moderation.status == 'allow'):
+            group_message.action_accept()
+
+        elif moderation and moderation.status == 'ban':
+            group_message.action_discard()
+
+        elif self.moderation and self.moderation_notify:
+            self.env['mail.mail'].sudo().create({
+                'author_id': self.env.user.partner_id.id,
+                'email_from': self.env.user.company_id.catchall_formatted or self.env.user.company_id.email_formatted,
+                'body_html': group_message.mail_group_id.moderation_notify_msg,
+                'subject': 'Re: %s' % (kwargs.get('subject', '')),
+                'email_to': kwargs.get('email_from'),
+                'auto_delete': True,
+                'state': 'outgoing'
+            })
+
+    # ------------------------------------------------------------
+    # MODERATION
+    # ------------------------------------------------------------
+
+    def _join_group(self, email, partner_id=None):
+        self.ensure_one()
+
+        if partner_id:
+            partner = self.env['res.partner'].browse(partner_id).exists()
+            email = partner.email
+
+        email_normalized = email_normalize(email)
+
+        existing_member = self.member_ids.filtered_domain([('email_normalized', '=', email_normalized)])
```

Comment:
We may have 10K+ members on some groups ... maybe a search could be easier to perform.